File: java/src/main/java/com/sasteval/vuln/ailegacy/NullCheckBypass.java
package com.sasteval.vuln.ailegacy;

import com.sasteval.util.DbUtil;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;

/**
 * CWE-754: Improper Check for Exceptional Conditions
 * VULNERABILITY: Null handling fails open and reaches an admin-only delete operation.
 */
public class NullCheckBypass extends HttpServlet {

    private boolean canDeleteUser(String token) {
        try {
            int prefixLength = token.length();
            return prefixLength > 10 && token.startsWith("admin-");
        } catch (NullPointerException e) {
            // VULN: Missing token unexpectedly grants access instead of denying it.
            return true;
        }
    }

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String token = request.getHeader("X-Admin-Token");
        String userId = request.getParameter("userId");
        response.setContentType("application/json");

        if (!canDeleteUser(token)) {
            response.setStatus(HttpServletResponse.SC_FORBIDDEN);
            response.getWriter().println("{\"error\": \"Admin token required\"}");
            return;
        }

        try (Connection conn = DbUtil.getConnection();
             PreparedStatement ps = conn.prepareStatement("DELETE FROM users WHERE id = ?")) {
            ps.setString(1, userId);
            int rows = ps.executeUpdate();
            response.getWriter().println("{\"deleted\": " + rows + "}");
        } catch (Exception e) {
            response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            response.getWriter().println("{\"error\": \"" + e.getMessage() + "\"}");
        }
    }
}

File: dotnet/Vuln/AiLegacy/NullCheckBypass.cs
using System.Data.SQLite;
using SastEval.Util;

namespace SastEval.Vuln.AiLegacy;

/// <summary>
/// CWE-754: Improper Check for Exceptional Conditions - Null handling fails open and
/// reaches an admin-only delete sink.
/// </summary>
public static class NullCheckBypass
{
    private static bool CanDeleteUser(string? token)
    {
        try
        {
            var length = token.Length;
            return length > 10 && token.StartsWith("admin-");
        }
        catch (NullReferenceException)
        {
            // VULNERABLE: Missing token unexpectedly grants access instead of denying it.
            return true;
        }
    }

    public static IResult Handle(HttpContext context)
    {
        var token = context.Request.Headers["X-Admin-Token"].ToString();
        var userId = context.Request.Query["userId"].ToString();

        if (!CanDeleteUser(string.IsNullOrEmpty(token) ? null : token))
        {
            return Results.Unauthorized();
        }

        var conn = DbUtil.GetConnection();
        using var cmd = new SQLiteCommand("DELETE FROM users WHERE id = @id", conn);
        cmd.Parameters.AddWithValue("@id", userId);
        var rows = cmd.ExecuteNonQuery();
        return Results.Ok(new { RowsAffected = rows });
    }
}
